src/expression_tree.rs

```diff
@@ -129,24 +99,290 @@ impl Expression {
     }
 
     //evaluates, then uses the approximation modifier on the tree
-    pub fn approximate<A: Modifier, E: Modifier, S: Modifier>(&self, approximator: &A, evaluator: &E, simplifier: &S) -> Result<f32, ()> {
+    pub fn approximate<A: Modifier, E: Modifier, S: Modifier>(&self, approximator: &A, evaluator: &E, simplifier: &S) -> Result<Numeric, ()> {
         let mut expr = self.clone();
 
         loop {
```

Comment:
Might be cleaner as
```rust
while approximator.modify(&mut expr) {
  /* ... */
}
```

Reply:
it would be cleaner that way, but unfortunately, the expression needs to be simplified before it's approximated.

Reply:
Oh ok, that makes sense.